datadog_checks_base/datadog_checks/base/checks/base.py

```diff
@@ -867,6 +867,16 @@ def check(self, instance):
         # type: (InstanceType) -> None
         raise NotImplementedError
 
+    def cancel(self):
+        # type: () -> None
+        """
+        This method is called when the check in unscheduled by the agent. This
+        is SIGNAL that the check is being unscheduled and can be called while
+        the check is running. It's up to the python implementation to make sure
+        cancel is thread safe and won't block.
```

Comment:
> won't block

Assume a check is running a thread in the background to offload some work.

What would `.cancel()` do in this case? From the wording above I understand it should take some action to ensure the thread terminates eventually (eg setting a `threading.Event`), but we _shouldn't_ call `.join()` because that would block, right?

Reply:
Exactly. The idea is to let the check know that it's been unscheduled and should terminate thread or free resources shared between instances. In short: since `cancel` might be called it's up to the check to not lock the GIL from both `cancel` and `run`.